docs_src/design/adr/security/0028-authentication.md

```diff
@@ -0,0 +1,272 @@
+
+* As an added bonus, Vault supports longer JWT key sizes than the Kong JWT plugin.
+
+Additionally, `security-bootstrapper` will need to modified to not
+block on availability of Postgres before issuing the ready-to-run signal.
+(This change is already completed.)
+
+
+### High-level list of changes
+
+The following list of changes is derived from the proof of concept implementation
+to actually effect the change (besides the prerequisite changes above):
+
+- Kong and Postgres is removed from compose files and snaps.
+
+- Add an NGINX reverse proxy with using the proxy auth module.
+
+- Create a new `security-proxy-auth` service to check the incoming JWT for validity.
+  (NGINX will be configured to delegate to this service for authentication checks.
+  NGINX could also delegate to a minimal function like /api/v2/version,
+  but the reason as to why the function was called wouldn't be as clear as
+  having a separate authentication service.)
+
+- The `security-proxy-setup` container remains, with the binary replaced
+  with a small shell script to create a default TLS certificate and key.
+
+- The `secrets-config` utility will create new users in Vault instead of Kong,
+  and update TLS configuration for NGINX on disk instead of the Kong API.
+
+- Modifications to `go-mod-core-contracts` to support an
+  injectable authentication interface to add JWT's to outgoing HTTP requests.
+
+- Modifications to `go-mod-bootstrap` to realize the `go-mod-secrets` changes,
+  create common JWT authentication handlers,
+  and inject JWT authentication to the core-contracts clients.
+
+- Modifications to individual EdgeX services to authenticate selected routes
+  (that is, every route except `/api/v2/ping`, which remains anonymous).
+
+- Modifications to `security-bootstrapper` to build an entrypoint script for
+  NGINX and a default NGINX configuration.
+
+- Documentation updates.
+
+
+## Decision
+
+Token-based authentication is flexible and works in a wide variety of use cases,
+but does not address issues of network security.
+
+For scenarios where all EdgeX services are running on the same host,
+or there is an existing solution to network security already in place,
+such as an encrypted network overlay
+as might be found in some Kubernetes deployments of EdgeX,
+the token-based solution offers significant
+memory and disk savings over the Kong-based solution
+used in EdgeX releases prior to 3.0.
+
+For scenarios where token-based authentication credentials can be exposed over a network,
+an authentication solution based on end-to-end encryption would be more appropriate.
+
+
+## Considerations
+
+### Size and Space Impact of Kong + Postgres Versus Alternatives
+
+#### Disk space
+
+Kong + Postgres (docker images): 353MB
+
+Alternatives:
+ - HAProxy: 40 MB
+ - NGINX: 12 MB - 41 MB depending on image selection (12 MB works for us)
+
+#### Memory
+
+Kong + Postgres: 160 MB (at startup)
+
+Alternatives:
+  - HAProxy: 80 MB
+  - NGINX: 5 MB
```

Comment:
This information is incomplete without specifying which versions of those applications and which Docker images were used. A different version and docker image base of those applications can have completely different resource usage.

Reply:
Specifics:

#### Disk space

A savings of up to ~300 MB in docker images can be expected,
depending on specific selection of container images used.
(The POC implementation successfully used the smallest NGINX
available, alpine-slim.)

| Image                            | Tag         | Image ID     | Age          | Size   |
|----------------------------------|-------------|--------------|--------------|--------|
| nginx                            | alpine      | 2bc7edbc3cf2 | 6 days ago   | 40.7MB |
| nginx                            | alpine-slim | c59097225492 | 6 days ago   | 11.5MB |
| nginx                            | latest      | 3f8a00f137a0 | 8 days ago   | 142MB  |
| kong                             | 2.8         | 0affcb95d383 | 6 days ago   | 139MB  |
| postgres                         | 13.8-alpine | 551b13d106b4 | 4 months ago | 213MB  |
| edgexfoundry/security-proxy-auth | 0.0.0-dev   | b2ee5c21efba | 8 days ago   | 16.2MB |


#### Memory

A memory savings of up to ~150 MB has been observed in the POC implementation
upon initial startup of the framework.

| CONTAINER ID | NAME            | CPU % | MEM USAGE / LIMIT   | MEM % | NET I/O         | BLOCK I/O       | PIDS |
|--------------|-----------------|-------|---------------------|-------|-----------------|-----------------|------|
| cad71e71ab32 | edgex-kong      | 0.03% | 109.4MiB / 15.61GiB | 0.68% | 255kB / 263kB   | 0B / 69.6kB     | 2    |
| 9ab4de1e5448 | edgex-kong-db   | 0.11% | 64.51MiB / 15.61GiB | 0.40% | 232kB / 183kB   | 32.2MB / 53.9MB | 18   |
| ff1e97c16e55 | edgex-nginx     | 0.00% | 4.289MiB / 15.61GiB | 0.03% | 3.24kB / 248B   | 0B / 0B         | 5    |
| 42629157e65c | edgex-proxyauth | 0.00% | 6.258MiB / 15.61GiB | 0.04% | 22.9kB / 16.2kB | 7.3MB / 0B      | 11   |

Reply:
Looks great. Much more detailed than I asked for. 
The only issue is the relative images ages without any anchor date.